macleann/enchiridion-client
# Fix #14 Get Tailwind up and running

# Major styling and syntax changes

Get ready folks, this one's a doozy. Most of it is simply editing the classNames in the JSX of each component's return, but there are some bits of new algorithmic changes that need to be mentioned:

### First and foremost, Tailwind:
- Added Tailwind and eventually DaisyUI dependencies. I'll mention this below, but anyone who has already followed the setup instructions will need to run `npm install` in their `the-enchiridion` directory  to see the benefits of what this PR has to offer.
- Added tailwind's output stylesheet to `index.html`
- Added custom Tailwind classes to `index.css`
- Added DaisyUI plugin to Tailwind's `tailwind.config.js` file

### Auth-related changes:
- Removed the default user in state on the login form 
- Changed the form placeholders to something more apt to the project in both the Register and Login components
- Changed the disabling button ternary logic and modified the classNames in both Register and Login components

### New additions:
- Added an `svgs` directory
- Added `Loading` and `TrashIcon` components to new `svgs` directory that return svg icons/animations
- Imported the `Loading` component and called it as the return for if the page was still fetching needed objects for state to the following components: `EpisodeDetail.js`, `PlaylistDetail.js`, `PlaylistForm.js`, `Playlists.js`, `SeasonDetail.js`, and `Seasons.js`

### Mobile-friendly changes:
- Changed classNames in `NavBar.js` to make it more mobile-friendly
- Added a function, useState, and useEffect that work in tandem to track the window's width and height at all times to `PlaylistForm.js`, `Playlists.js`, and `Seasons.js`.
- Also added ternary logic in each of those components' return statements to render the page differently depending on whether the user was on mobile or not (768 pixel width breakpoints).

### Major Changes:

`PlaylistForm.js`:
- The episode dropdown is hidden until the user has selected a season
- The playlist's added episodes render with the still and overview on desktop, but not on mobile
- Changed a copy of episodes to add a `series_id` property to each episode instead of changing the response's list of episodes
- Added the same focusOrFilled functions as in `Login.js` and `Register.js`
- Added a function that triggers a window alert if any of the required playlist fields aren't filled out
- Changed classNames on the form to make it more mobile-friendly
- Added ternary logic to the "Save Playlist" button to disable it if any required fields were missing.

`Playlists.js`:
- Added a function that calculate total runtime of any given playlist
- Added a function that returns a grid of 4 episode images so long as the playlist has at least 1 episode that will serve as the playlist's image
- Added ternary logic to the return statement that shows each playlist differently depending on whether the user is on mobile or not

### Minor changes:
- Changed classNames in `PlaylistDetail.js`
- Changed classNames in `SeasonDetail.js`

### Known bugs:
- Removing an episode from a playlist completely breaks all Tailwind styling

### Bug fixes:

Fixes #14 Get Tailwind up and running

## Type of change


# How to Test?

Make sure you've followed all installation and setup instructions in the README.

```
git fetch origin nm-tailwind
git checkout nm-tailwind
npm install
npm start
```

- [ ] Test logging out, logging in, registering a new user
- [ ] Go to playlists, test adjusting the window width (I think this looks really good!)
- [ ] Go to create playlists, test adjusting window width
- [ ] Add some episodes, adjust the window size width
- [ ] Go to seasons, test adjusting the window width

# Checklist:
